dkeyserver/Enclave/enclave.cpp

```diff
@@ -97,30 +226,270 @@ sgx_status_t sgx_get_domainkey(uint8_t *domain_key)
+    // create a new SSL structure for a connection
+    if ((ssl_session = SSL_new(handler_ctx.ssl_server_ctx)) == nullptr)
+    {
+        log_d(TLS_SERVER
+              "Unable to create a new SSL connection state object\n");
+        goto exit;
+    }
 
-    if (p_dst_len < SGX_DOMAIN_KEY_SIZE)
-        return SGX_ERROR_UNEXPECTED;
+    if (SSL_set_fd(ssl_session, handler_ctx.client_socket_fd) != 1)
+    {
+        log_d(TLS_SERVER
+              "SSL set fd failed\n");
+        goto exit;
+    }
 
-    sgx_status_t ret = sgx_get_domainkey(domain_key);
-    if (ret != SGX_SUCCESS) {
-        printf("Failed to get domain:%d.\n", ret);
-        return ret;
+    // wait for a TLS/SSL client to initiate a TLS/SSL handshake
+    log_i(TLS_SERVER "initiating a passive connect SSL_accept\n");
+    test_error = SSL_accept(ssl_session);
+    if (test_error <= 0)
+    {
+        log_d(TLS_SERVER " SSL handshake failed, error(%d)(%d)\n",
+              test_error, SSL_get_error(ssl_session, test_error));
+        goto exit;
+    }
+
+    log_d(TLS_SERVER "<---- Read from client:\n");
+    if (read_from_session_peer(
+            ssl_session, CLIENT_PAYLOAD, CLIENT_PAYLOAD_SIZE) != 0)
+    {
+        log_d(TLS_SERVER " Read from client failed\n");
+        goto exit;
+    }
+
+    for (unsigned long int i = 0; i < SGX_DOMAIN_KEY_SIZE; i++)
+    {
+        log_d("domain_key[%u]=%2u\n", i, handler_ctx.domainkey[i]);
+    }
+
+    log_d(TLS_SERVER "<---- Write to client:\n");
+    if (write_to_session_peer(
+            ssl_session, handler_ctx.domainkey, SGX_DOMAIN_KEY_SIZE) != 0)
+    {
+        log_d(TLS_SERVER " Write to client failed\n");
+        goto exit;
+    }
+    if (handler_ctx.client_socket_fd > 0)
+    {
+        ocall_close(&ret, handler_ctx.client_socket_fd);
+        if (ret != 0)
+        {
+            log_d(TLS_SERVER "OCALL: error closing client socket before starting a new TLS session.\n");
+            goto exit;
+        }
+    }
+exit:
+    SSL_free(ssl_session);
+    log_i("write domainkey to clent success\n");
+    return ((void *)0);
+}
+
+int create_listener_socket(int port, int &server_socket)
+{
+    int ret = -1;
+    const int reuse = 1;
+    struct sockaddr_in addr;
+    addr.sin_family = AF_INET;
+    addr.sin_port = htons(port);
+    addr.sin_addr.s_addr = htonl(INADDR_ANY);
+
+    server_socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (server_socket < 0)
+    {
+        log_d(TLS_SERVER "socket creation failed\n");
+        goto exit;
+    }
+
+    if (setsockopt(
+            server_socket,
+            SOL_SOCKET,
+            SO_REUSEADDR,
+            (const void *)&reuse,
+            sizeof(reuse)) < 0)
+    {
+        log_d(TLS_SERVER "setsocket failed \n");
+        goto exit;
+    }
+
+    if (bind(server_socket, (struct sockaddr *)&addr, sizeof(addr)) < 0)
+    {
+        log_d(TLS_SERVER "Unable to bind socket to the port\n");
+        goto exit;
+    }
+
+    if (listen(server_socket, 20) < 0)
```

Comment:
define the max accept number as a Marco, e.g. CLIENT_MAX_NUM

Reply:
it will be changed

Reply:
ok refined